doc/Language/iterating.pod6

```diff
@@ -4,136 +4,321 @@
+Keeping the aforementioned caveats in mind, implementing this low-level
+interface simplifies the implementation of the C<Iterable> interface. Now the
+iterator will be the object itself, since we can invoke C<pull-one> on it to
+access every member in turn; C<iterator> will thus return just C<self>, which is
+possible since the object will be both an C<Iterable> and an C<Iterator>.
+
+This need not always be the case, and in most cases C<iterator> will have to
 build an iterator type to be returned (that will, for instance, keep track of
 the iteration state, which we are doing now in the main class), such as we did
-in the previous example; however, this example shows the minimal code needed to
-build a class that fulfills the iterator and iterable roles.
+in the previous section; however, this example shows the minimal code needed to
+build a class that fulfills both the iterator and iterable roles.
+
+=head1 Manual iteration
+
+Iterating constructs, such as C<for> loops, use iterators to iterate through
+collections. However, the programmer can also request values from an iterator
+one at a time without the aid of any iterating construct.
+
+You can get an iterator from any iterable, and you can use an iterator to manually
+loop over the iterable it came from. Thus, the way to manually iterate through an
+iterable is to get a hold of an iterator object by invoking the C<iterator> method
+and call the C<pull-one> method on said iterator to request a new item until
+C<IterationEnd> is returned. 
+
+=begin code
+my @languages = <perl python haskell ruby>;
+my $iter = @languages.iterator;
+
+say $iter.pull-one; #=> «perl␤» 
+say $iter.pull-one; #=> «python␤» 
+say $iter.pull-one; #=> «haskell␤» 
+say $iter.pull-one; #=> «ruby␤» 
+say $iter.pull-one; #=> «IterationEnd␤» 
+=end code
+
+After all the iterator's values are consumed, the returned value is the constant
+C<IterationEnd>. Once C<IterationEnd> has been generated, it's forbidden to make
+attempts to fetch more data, and behavior for doing so is undefined. The only
+valid use of the sentinel value C<IterationEnd> in a program is identity
+comparison (using L<C<=:=>|/routine/=:=>) with the result of a method in the
+iterator API. Any other behavior is undefined and implementation dependent.
+
+Thus, the manual iteration can be rewritten as a loop:
+
+=begin code
+my @languages = <perl python haskell ruby>;
+
+my $iter = @languages.iterator;
+loop {
+    my \value = $iter.pull-one;
+    last if value =:= IterationEnd;
+    say value;
+}
+=end code
+
+Since C<IterationEnd> is a constant, it's important that the variable we compare
+against it be bound, and not assigned. Hence, the use of the sigilless variable
+C<value> in the previous example.
 
 =head1 How to iterate: contextualizing and topic variables
 
-C<for> and other loops place the item produced in every iteration into the
+Loops, such as C<for>, place the item produced in every iteration into the
```

Comment:
"Loops" don't do that, so this turn of the sentence is not totally correct. For and others do. "loop" doesn't place the item in the topic variable, or while IARC.